packages/flutter/lib/src/widgets/async.dart

```diff
@@ -243,26 +266,20 @@ class AsyncSnapshot<T> {
 
   /// Returns a snapshot like this one, but in the specified [state].
   ///
-  /// The [data] and [error] fields persist unmodified, even if the new state is
-  /// [ConnectionState.none].
-  AsyncSnapshot<T> inState(ConnectionState state) => AsyncSnapshot<T>._(state, data, error);
-
-  /// Returns whether this snapshot contains a non-null [data] value.
-  ///
-  /// This can be false even when the asynchronous computation has completed
-  /// successfully, if the computation did not return a non-null value. For
-  /// example, a [Future<void>] will complete with the null value even if it
-  /// completes successfully.
-  bool get hasData => data != null;
+  /// The [hasData], [data], [hasError], and [error] fields persist unmodified,
+  /// even if the new state is [ConnectionState.none].
+  AsyncSnapshot<T> inState(ConnectionState state) => AsyncSnapshot<T>._(state, hasData, _data, error);
 
   /// Returns whether this snapshot contains a non-null [error] value.
   ///
   /// This is always true if the asynchronous computation's last result was
   /// failure.
+  ///
+  /// When this is true, [hasData] will always be false.
   bool get hasError => error != null;
 
   @override
-  String toString() => '$runtimeType($connectionState, $data, $error)';
+  String toString() => '$runtimeType($connectionState, $hasData, $_data, $error)';
```

Comment:
hasData is a bool, so that will be confusing. maybe `${hasData ? "has data" : "no data"}` ?

Reply:
also, probably omit `$_data` if there's no data, and omit error if there's no error? in general this could be clearer in various ways.